kore/src/SMT.hs

```diff
@@ -415,19 +400,23 @@ the 'Solver' is never returned to the 'MVar', so any threads waiting for the
 solver will hang.
 
  -}
-stopSolver :: MVar Solver -> IO ()
-stopSolver mvar = do
-    solver <- takeMVar mvar
+stopSolver :: Logger -> MVar SolverHandle -> IO ()
+stopSolver logger mvar = do
+    solverHandle <- takeMVar mvar
+    let solver = Solver solverHandle logger
     _ <- SimpleSMT.stop solver
     return ()
 
 -- | Run an external SMT solver.
 runSMT :: Config -> Logger -> SMT a -> IO a
-runSMT config logger SmtT { runSmtT } =
+runSMT config logger SMT { getSMT } =
     Exception.bracket
-        (newSolver config logger `Exception.catch` showZ3NotFound)
-        stopSolver
-        (runReaderT runSmtT)
+        ( catch
+            (newSolver config logger)
+            showZ3NotFound
+        )
+        (stopSolver logger)
+        (\mvar -> runReaderT (getLoggerT $ runReaderT getSMT mvar) logger)
```

Comment:
At this place, and in `withSolver`, it looks like we would benefit from a `runSMT' :: MVar SolverHandle -> Logger -> SMT a -> IO a`.